text/0000-read-only-build-layers.md

```diff
@@ -0,0 +1,98 @@
+2. Buildpack #2 uses the python binary to invoke pip to provide further dependencies.
+
+Unbeknownst to the buildpack author, python has a side-effect on execution where if it's installation directory is writable, it creates new cache objects or updates existing files. As a result, execution of buildpack #2 causes the content of the layers provided by buildpack #1 to change, causing the final layer imageID to change and hampers layer reproducibility and new layers have to be pushed out to the registry.
+
+This could easily be fixed if buildpack #1 only exposed its build layers as read-only to subsequent buildpacks or if it discarded the modifications made by subsequent buildpacks essentially making it "read-only" in the context of the build process.
+
+## What is the expected outcome?
+
+The spec and lifecycle is modified to support the above use cases.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Updates to the lifecycle so that `build` layers created by buildpacks should be layerized immediately after the buildpack that provided the layer is finished with its build process. This way any modification made by the future buildpacks would be ignore while still making this layer available to future buildpacks.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+The lifecycle will have to layerize the layers marked `build` as `true` during the `build` phase of the lifecycle instead of the `export` phase.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Possibly mixing the concerns of the `build` and the `export` phases.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered?
+
+We could change the the layer so that it is owned by root or a different user between the build steps of different buildpacks so that subsequent buildpacks cannot modify the layers created by other buildpacks. This would require elevated privileges during the build phase of the lifecycle.
+
+
+- Why is this proposal the best?
+
+This proposal allows you to achieve the above use cases without changing the expected UID/GID of the layers and without elevated permissions during the build phase.
+
+- What is the impact of not doing this?
+
+We could have unknown layer modifications by buildpacks which is currently a spec violation.
+
+# Prior Art
+[prior-art]: #prior-art
+
+TODO
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+The impact this would have on the performance of the export step. The exact details of which parts of the lifecycle build and export phases that need to be changed.
+
+How to handle the cases when users want a common build workspace that is cached. This will be covered in a subsequent RFC and that RFC and this RFC should be implemented as an "atomic" change to avoid existing use cases from being broken without a proper migration path.
+
+# Spec. Changes (OPTIONAL)
+[spec-changes]: #spec-changes
+
+None.
```

Comment:
Another thing to consider is the calculation of the layer diffIDs. We currently do this in parallel while archiving avoiding the need to read the layer data from disk a second time. To avoid degrading the performance of the build we will need to move this calculation to the build phase and record the results for the exporter to use.

Reply:
Updated the implementation details slightly. The `builder` and `exporter` arguments would remain the same and the `builder` would just output layerized tars to `/layers/exported` which the `exporter` could read from. This would avoid a platform api change. We could re-evaulate this as a part of #145 and change it in the future to be explicitly passed but I would prefer that older platform apis with the newer lifecycle also get this behaviour.